### ISSUE TITLE
Docker not shutting down gracefully on reboot

### ISSUE THREAD

**Issue by Saulis** [coreos/bugs#243]
This issue is related to #186 but it also happens without defining unit files in systemd.

I've tested this in AWS using CoreOS-stable-522.4.0-hvm (ami-8f0087f8) and CoreOS-beta-557.0.0 (ami-1d6dec6a).

Having a really simple container:
```
FROM debian:wheezy

COPY trap.sh /trap.sh
ENTRYPOINT ["/trap.sh"]
```
trap.sh simulates process which takes 5 seconds to shutdown.
```
#!/bin/bash

trap "echo Exiting && sleep 5 && exit" SIGINT SIGTERM
while :
do
  echo "Working..."
  sleep 2
done
```

When I start and stop the container manually, the container waits 5 seconds until it is stopped:
```
#from CoreOS 522.4.0
docker start trap
docker stop trap
journalctl -u docker -e

Jan 16 07:15:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [info] POST /v1.15/containers/trap/start
Jan 16 07:15:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job start(trap)
Jan 16 07:15:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job allocate_interface(54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233)
Jan 16 07:15:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job allocate_interface(54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233) = OK (0)
Jan 16 07:15:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job log(start, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest)
Jan 16 07:15:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job log(start, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest) = OK (0)
Jan 16 07:15:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job start(trap) = OK (0)
Jan 16 07:15:10 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [info] POST /v1.15/containers/trap/stop?t=10
Jan 16 07:15:10 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job stop(trap)
Jan 16 07:15:17 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job log(die, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest)
Jan 16 07:15:17 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job log(die, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest) = OK (0)
Jan 16 07:15:17 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job release_interface(54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233)
Jan 16 07:15:17 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job release_interface(54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233) = OK (0)
Jan 16 07:15:17 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job log(stop, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest)
Jan 16 07:15:17 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job log(stop, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest) = OK (0)
Jan 16 07:15:17 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job stop(trap) = OK (0)
```
Now, if I start the container and then reboot, docker is shut down without taking the time to shutdown the container gracefully:
```
docker start trap
sudo reboot
#after reboot
journalctl -u docker -e
Jan 16 07:17:50 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [info] POST /v1.15/containers/trap/start
Jan 16 07:17:50 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job start(trap)
Jan 16 07:17:50 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job allocate_interface(54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233)
Jan 16 07:17:50 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job allocate_interface(54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233) = OK (0)
Jan 16 07:17:50 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] +job log(start, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest)
Jan 16 07:17:50 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job log(start, 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233, trap:latest) = OK (0)
Jan 16 07:17:50 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [72d50836] -job start(trap) = OK (0)
Jan 16 07:18:04 ip-172-31-47-43.eu-west-1.compute.internal systemd[1]: Stopping Docker Application Container Engine...
Jan 16 07:18:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: 2015/01/16 07:18:04 Received signal 'terminated', starting shutdown of docker...
Jan 16 07:18:04 ip-172-31-47-43.eu-west-1.compute.internal docker[666]: [error] container.go:186 Error logging event die for 54db2e17a6535bb412293a295fb07b3539a56ec1bcec9bc2b10844104c186233: engine is shutdown
Jan 16 07:18:04 ip-172-31-47-43.eu-west-1.compute.internal systemd[1]: Stopped Docker Application Container Engine.
-- Reboot --
```
Sometimes you can also see those logging event errors because the container engine has already been shutdown.

I would be happy to hear if anyone would have at least any workaround for this issue.

**Comment by marineam** [comment]
You don't need to be using systemd services in order to hit https://github.com/coreos/bugs/issues/186, it is a flaw in the interaction between docker and systemd when creating the cgroup for containers.

**Comment by marineam** [comment]
And I don't know of a workaround beyond manually stopping containers (or docker) before stopping the system.

**Comment by Saulis** [comment]
Do you know if there's a way I could run commands with systemd before it starts shutting down docker?

**Comment by crawford** [comment]
Fixed in 584.0.0.
